### ISSUE TITLE
PVImage wx python gui object

### ISSUE THREAD

**Issue by vstadnytskyi** [pyepics/pyepics#184]
I am trying to create an object that will allow me to create bitmaps and import it into my GUI so I can visualize images from a camera transmitted with CA. I called it [PVImage](https://github.com/vstadnytskyi/pyepics/blob/cfb05fd2a77ebf02855c566cb655df4c97837751/epics/wx/wxlib.py#L462) to follow the pattern of other GUI objects in pyepics(e.g. PVStaticText. You can see also see the code at the bottom.

I have created two files: [io](https://github.com/vstadnytskyi/caproto-sandbox/blob/master/caproto_sandbox/image_io.py) and https://github.com/vstadnytskyi/caproto-sandbox/blob/master/caproto_sandbox/image_gui.py

caproto server
- image(3x100x100)  actual image
- image_mean - mean value of the image.
The image_mean was created to see if anything is updating on the screen.

Here is how GUI looks.
<img width="296" alt="image" src="https://user-images.githubusercontent.com/29989124/68984314-01ed9f80-07d5-11ea-9e75-0211c31f473f.png">

the problem is that the image is not updating. I had one instance when image was updating at 20 Hz. I was super excited. And after that, I cannot make it work again. I have put print statements to see what is going on and the function OnPVChange is getting called only once at the very beginning. It is supposed to be a callback function and should be called every time the new PV arrives. On the other hand, the image_mean value is updating on a clock as expected.

The PVImage
``` python
class PVImage(wx.StaticBitmap, PVMixin):
    """ Static text for displaying a PV value,
        with callback for automatic updates
        By default the text colour will change on alarm states.
        This can be overriden or disabled as constructor
        parameters
        """
    def __init__(self, parent, pv=None, style=None, im_size = None,
                 minor_alarm="DARKRED", major_alarm="RED",
                 invalid_alarm="ORANGERED", **kw):
        self.im_size = im_size
        wstyle = wx.ALIGN_LEFT
        if style is not None:
            wstyle = style

        wx.StaticBitmap.__init__(self, parent, wx.ID_ANY,
                               style=wstyle, **kw)
        PVMixin.__init__(self, pv=pv)
        self._fg_colour_alarms = {
            epics.MINOR_ALARM : minor_alarm,
            epics.MAJOR_ALARM : major_alarm,
            epics.INVALID_ALARM : invalid_alarm }

    def _SetValue(self, value):
        "set widget label"
        print('_SetValue value = ',value)
        from PIL import Image
        import io
        if value is not None:
            # I do all the stuff below to convert the array to the right format. It seems to be transmitted as 32bit instead of 8, even though i tried created it as 8 on the server. This needs to be investigated.
            raw_image = value
            im_mode = 'RGB'
            im_size = self.im_size
            img = Image.frombuffer(im_mode, im_size, raw_image,
                                    'raw', im_mode, 0, 1)
            buf = io.BytesIO()
            img.save(buf, format='PNG')
            buf.seek(0)
            image = Image.open(buf)
            width, height = image.size
            bit_img = wx.Bitmap.FromBuffer(width, height, image.tobytes())
            self.SetBitmap(bit_img)
        else:
            print('_SetValue',value)

    @EpicsFunction
    def OnPVChange(self, value):
        "called by PV callback"
        import traceback
        from numpy import array
        try:
            print('Try OnPVChange')
            print('self.pv = ',self.pv)
            print('self.pv.get()',self.pv.get())
            print('value OnPVChange = ',value,len(value),type(value))
            # I don't know how to work with value correctly.
            # the value of value is <array size=30000, type=ctrl_long>
            # the type of value is <class 'str'>
            if self.pv is not None:
                value = array(self.pv.get(), dtype = 'int8')
                self._SetValue(value)
```

**Comment by vstadnytskyi** [comment]
I found something:
Everything works if my caproto server creates PV this way:
`image = pvproperty(value=arr, dtype = int)`
But doesn't work if it is created this way
`image = pvproperty(value=arr, dtype = int, max_length = 4000000)`

Somehow presence of max_length screws up the initial setup of pyepics GUI. However, it works if I initialize GUI with the server that doesn't have max_length and later restart it making a change in the code to include max_length.

also, I cannot transmit image 500x500.

**Comment by vstadnytskyi** [comment]
I think I have resolved the issue. The maximum length of the PV allowed for automonitor is specified in [ca.py](https://github.com/pyepics/pyepics/blob/ad941d5b93a709f1cc9640a2b2828efda970b026/epics/ca.py#L71)and is 
`AUTOMONITOR_MAXLENGTH = 65536 # 16384`

**Comment by newville** [comment]
@vstadnytskyi you might have a look at the area Detector viewer at https://github.com/pyepics/epicsapps/tree/master/epicsapps/areadetector

This is all under-documented and poorly advertised, but many of the "epicsapps" are actually in active use (so, "somewhat debugged") at GSECARS -- areaDetector viewer and Instruments are probably the most widely useful.

The areaDetector viewer shows images from a basic monochrome area detector.  It supports a "spyglass", contrast levels, selecting color tables, and it is easy to add other PVs to the windows.   You have to write a yaml file to configure it (see https://github.com/pyepics/epicsapps/blob/master/examples/areaDetector/ID13_Eiger500K.yaml for an example).  For a color camera (like for a visible light microscope), you can look at https://github.com/pyepics/epicsapps/tree/master/epicsapps/microscope.  That app is really about more than just viewing the image, but the code for converting a color area Detector image to a wxBitmap is there.   

I'd be happy to show these apps to you, or you can stop by 13ID-E to see the microscope display (there are often two of these running full time).   

For image data of any sizeable image, you will need to set EPICS_CA_MAX_ARRAY_BYTES on client and server side to get the image data transferred.  

Also -- yes, use areaDetector and a soft epics IOC.  Don't try to emulate this functionality by writing your own CA server.

**Comment by vstadnytskyi** [comment]
@newville following-up on this issue. Another reason you create this PVImage PVMixin object, is to be able to generate a figure with matplotlib and save it as bitmap on a server (CA server) and transfer the bitmap graph to a GUI for visualization. This simplifies programming of GUIs if there is a need to visualize some data(if the length and size is standard)

Here is an example of a GUI I made with PyEpics + PVImage
<img width="1971" alt="image" src="https://user-images.githubusercontent.com/29989124/129913905-a639c1c1-fc02-4f05-a9ed-81e12c071f7a.png">

@newville you have mentioned I should use https://github.com/pyepics/epicsapps/tree/master/epicsapps/areadetector to show images from a detector. 

How often does one want to visualize a graph without actually generating it on the GUI side?

**Comment by newville** [comment]
@vstadnytskyi I'm not very good at remembering partial issues from almost 2 years ago.  ;). 

For 2D line plots, I would normally want to send the data only over the wire and let the client application display that.  For wxPython apps, I would (and do!) use `wxmplot` (but I'm biased, as I wrote this).  For web apps I would (and do!) use plotly.  Both of these are pretty simple to make a basic 2D plot, and both give the end-user the ability to zoom in/zoom out, toggle log/linear scales, and (for wxmplot) change color schemes and font sizes.  If you send a bitmap, all of those choices are locked in when the bitmap is created.

For areaDetector viewing, I typically use the epicsapps areadetector display app (then again, I wrote this, so biased).   This takes a YAML file configuration, but does allow selecting colormaps, auto-setting contrast levels (which can be very useful for cameras with bad or hot pixels or XRD patterns), and adding selected PVs for controlling the camera (say, AcquireTime, Triggering Mode, etc).   The guts of "display new image from camera" is in https://github.com/pyepics/epicsapps/blob/master/epicsapps/areadetector/imagepanel.py

This sets a monitor on `cam1:ArrayCounter_RBV` that causes a wx Refresh event.  That "Paint method" gets the data as a numpy array, converts that to a wx.Image.  That image will be scaled (in x,y size) for the display, maybe rotated or flipped, a selected colormap applied, and perhaps rescaled in intensity according to contrast selection (by percentiles of the raw image).  That wx.Image is then drawn to the screen (convert to wx.Bitmap using a device context to paint to the screen (double-buffered to minimize flicker).

For sure, there are other choices for displaying data from Epics PVs.  Basically, I have tools that work for us and am happy to help others use them.  I have not seen anything that is even remotely as good as these tools for displaying Epics array data, but I'm not really claiming these are useful for everyone.

**Comment by vstadnytskyi** [comment]
> @vstadnytskyi I'm not very good at remembering partial issues from almost 2 years ago. ;).

@newville Yeah. I know. It has been a while. I didn't realize it has been almost 2 years.

This issues is about an extra PVMixin object I called PVImage. Pretty much GUI object that shows image from a PV. Both client and server need to agree what is the size of the image and how it is encoded (e.g. RGB 8bit). Currently, I am using it to show a graph that I plot on a server side. This was done to simplify coding on the GUI side. Another solution is to plot on a GUI side which I do not know how to do easily. I will look into 'wxmplot' library. You need less bandwidth to transfer X-Y data and plot on GUI side. But it will take computational power to make plot on the GUI side. I had issues before that my GUI would freeze for a moment when graph is rendered.

**Comment by newville** [comment]
@vstadnytskyi 

> This issues is about an extra PVMixin object I called PVImage(https://github.com/vstadnytskyi/icarus-nmr/blob/d9fa6747582a6d2addf35b28ca339a4424eb26b3/icarus_nmr/pyepics.py). Pretty much GUI object that shows image from a PV

That's really more like a PV Bitmap - it derives from wx.Bitmap, not from wx.Image.  Currently, epics.wx.PVBitmap expects to select a bitmap from a finite set of values -- like maybe a set of colored images for an Enum PV or something.  

It might be reasonable to restructure that so that one could supply a method to render a bitmap based on the data value -- I think that is basically what you are doing.
